pyo3-macros/src/lib.rs

```diff
@@ -55,26 +67,152 @@ pub fn pyproto(_: TokenStream, input: TokenStream) -> TokenStream {
+/// | `weakref` | Allows this class to be [weakly referenceable][6]. |
+/// | `extends = BaseType`  | Use a custom baseclass. Defaults to [`PyAny`][4] |
+/// | `subclass` | Allows Python classes to inherit from this class.  |
+/// | `unsendable`<!-- hack to stop the column from wrapping -->&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160; | Required if your struct is not [`Send`][3]. Rather than using `unsendable`, consider implementing your struct in a threadsafe way by e.g. substituting [`Rc`][8] with [`Arc`][9]. By using `unsendable`, your class will panic when accessed by another thread.|
+/// | `module = "module_name"` |  Python code will see the class as being defined in this module. Defaults to `builtins`. |
+///
+/// For more on creating Python classes,
+/// see the [class section of the guide][1].
+///
+/// [1]: https://pyo3.rs/main/class.html
+/// [2]: https://pyo3.rs/main/class.html#customizing-the-class
+/// [3]: std::marker::Send
+/// [4]: ../prelude/struct.PyAny.html
+/// [5]: https://pyo3.rs/main/class/protocols.html#garbage-collector-integration
+/// [6]: https://docs.python.org/3/library/weakref.html
+/// [7]: https://doc.rust-lang.org/nomicon/leaking.html
+/// [8]: std::rc::Rc
+/// [9]: std::sync::Arc
+/// [10]: https://en.wikipedia.org/wiki/Free_list
 #[proc_macro_attribute]
 pub fn pyclass_with_inventory(attr: TokenStream, input: TokenStream) -> TokenStream {
     pyclass_impl(attr, input, PyClassMethodsType::Inventory)
 }
 
+/// A proc macro used to expose methods to Python.
+///
+/// Methods within a `#[pymethods]` block can be annotated with the following:
+///
+/// |  Annotation  |  Description |
+/// | :-  | :- |
+/// | [`#[new]`][4]  | Defines the class constructor, like Python's `__new__` method. |
+/// | [`#[getter]`][5] and [`#[setter]`][5] | These define getters and setters, similar to Python's `@property` decorator. This is useful for getters/setters that require computation or side effects; if that is not the case consider using [`#[pyo3(get, set)]`][11] on the struct's field(s).|
+/// | [`#[staticmethod]`][6]| Defines the method as a staticmethod, like Python's `@staticmethod` decorator.|
+/// | [`#[classmethod]`][7]  | Defines the method as a classmethod, like Python's `@classmethod` decorator.|
+/// | [`#[call]`][8]  | Allows Python code to call a class instance as a function, like Python's `__call__` method. |
+/// | [`#[classattr]`][9]  | Defines a class variable. |
+/// | [`#[args]`][10]  | Define a method's default arguments and allows the function to receive `*args` and `**kwargs`.  |
```

Comment:
There's actually also `#[name = "..."]`, but that's not documented. I am planning to move it to `#[pyo3(name = "...")]`, and same for `#[args]`. Then we can document all the `#[pyo3]` options. See #1567 - I'd be really interested in your opinion on this.

Individual arguments can also have `#[pyo3(from_py_with)]`.

Reply:
I think it's best to wait with making guides about `#pyo3(..)]` until that has happened.